test/AspNet.Security.OAuth.Providers.Tests/Shopify/ShopifyLoopbackRedirectHandler.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Net.Http;
+using System.Web;
+using AspNet.Security.OAuth.Infrastructure;
+
+namespace AspNet.Security.OAuth.Shopify
+{
+    internal class ShopifyLoopbackRedirectHandler : LoopbackRedirectHandler
+    {
+        private const string FormatShopParameter = "{0}.myshopify.com";
+
+        public ShopifyLoopbackRedirectHandler()
+        {
+        }
+
+        public string ShopName { get; set; }
+
+        protected override Uri BuildLoopbackUri(HttpResponseMessage responseMessage)
+        {
+            // Rewrite the URI to loop back to the redirected URL to simulate the user having
+            // successfully authenticated with the external login page they were redirected to.
+            var queryString = HttpUtility.ParseQueryString(responseMessage.Headers.Location.Query);
+
+            string location = queryString["redirect_uri"] ?? RedirectUri;
+            string state = queryString["state"];
+
+            var builder = new UriBuilder(location);
+
+            // Retain the _oauthstate parameter in redirect_uri for WeChat (see #262)
+            const string OAuthStateKey = "_oauthstate";
+            var redirectQuery = HttpUtility.ParseQueryString(builder.Query);
+            string oauthState = redirectQuery[OAuthStateKey];
+
+            // Remove any query string parameters we do not explictly need to retain
+            queryString.Clear();
+
+            queryString.Add("code", "a6ed8e7f-471f-44f1-903b-65946475f351");
+            queryString.Add("state", state);
+            queryString.Add("shop", String.Format(FormatShopParameter, ShopName));
```

Comment:
There's a lot of copy-paste here.

Maybe you could do this?

```csharp
protected override Uri BuildLoopbackUri(HttpResponseMessage responseMessage)
{
    Uri uri  = base.BuildLoopbackUri(responseMessage);

    var builder = new UriBuilder(uri);

    var queryString = HttpUtility.ParseQueryString(builder.Query);

    queryString.Add("shop", string.Format(FormatShopParameter, ShopName));

    builder.Query = queryString.ToString();

    return builder.Uri;
}
```

Reply:
Ok done